### ISSUE TITLE
Terraform not maintaining state in s3 backend.

### ISSUE THREAD

**Issue by cantide5ga** [hashicorp/terraform#15759]
**Terraform v0.10.0**

I reached out on gitter about this, but decided that in parallel to open an issue here.  If this isn't a bug, at the very least its a documentation enhancement.  Willing to help fix things here if necessary.

I've made some progress on this with an entry in my `main.tf`
```
terraform {
  backend "s3" {
    bucket = "mybucket"
    key    = "path/to/my/key/some.tfstate"
    region = "us-east-1"
  }
}
```
And a few hcl files, one for each environment containing:
```
bucket = "mybucket"
key    = "path/to/my/key"
region = "us-east-1"
```
And executing init with: `terraform init -backend-config some.hcl`

I also get a message of:
```
Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.
```

But there is no state file in my bucket - rather a `.terraform` folder locally with the usual stuff.

**Comment by jbardin** [comment]
Hi @cantide5ga,

Sorry, I'm not sure I understand what you're expecting to see.
Have you run any commands that write a state file? Running `terraform init` itself doesn't create a state file (though a backend implementation can choose to write whatever it wants, that is an implementation detail not part of the documented behavior), and running `init` with a local state won't write a state file either.

**Comment by cantide5ga** [comment]
Sorry for not being clear @jbardin - I'm running `terraform apply` after this initialization.

**Comment by jbardin** [comment]
Does running `terraform show` display the resources that you just applied? 
Are you running these commands from the same directory?
Terraform current stores the initialized backend in json in `.terraform/terraform.tfstate`. Can you show the `"backend"` block from that file?

Thanks!

**Comment by cantide5ga** [comment]
> Does running terraform show display the resources that you just applied?
Yes.

> Are you running these commands from the same directory?
Yes.

> Can you show the "backend" block from that file?
```
"backend": {
        "type": "s3",
        "config": {
            "bucket": "com-aquicore-terraform-state-dev",
            "key": "data-ingestion-tier",
            "region": "us-east-1"
        },
        "hash": 4870549691502983409
    }
```
Thanks for the help here.  Ultimate goal is having different state configurations as different buckets per environment.  I understand my example in the terraform block in original post has this hard coded, so perhaps there is something different that should be done in my case.  Just trying to understand this before diving into parameterizing.

**Comment by cantide5ga** [comment]
For some reason, it is there now.  Very annoying - sorry for the false alarm.

What is responsible for the state tracking configuration?  Is the terraform block in the top example required?  It's great to be able to declare a backend-config on init, but it is unclear how this all comes together to allow for parameterized backend configurations.

There isn't much to say on this use case on these two pages:
https://www.terraform.io/docs/backends/config.html
https://www.terraform.io/docs/backends/types/s3.html

Is there a better reference for this?

**Comment by jbardin** [comment]
No problem, glad you sorted it out. 

The `"terraform"` block is required, and is the only place where you can declare the `"backend"`.  As you have found, the parametrization of the backend is somewhat limited for a few reasons, because all the values for the backend must be present at `init` time, if they change later (e.g. through variables) it would invalidate the initialized config, and fundamentally because this needs to be loaded before we can run the interpolator to evaluate any variables. 

Right now you can can either delegate the separate environments to [workspaces](https://www.terraform.io/docs/state/workspaces.html), or you can carefully init again between changing environments using the `-reconfigure` flag to avoid the migration attempts. 

We definitely hope to improve this workflow in future releases.

Best,
James

**Comment by cantide5ga** [comment]
@jbardin 
After some more tinkering, I came upon the solution of only specifying a partial terraform backend block:
```
terraform {
  backend "s3" {}
}
```
and then parameterizing with `terraform init -backend-config dev.hcl` via an HCL file having 
```
bucket = "mybucket"
key    = "path/to/my/key"
region = "us-east-1"
```
Is there some caveat to doing this?  I actually think this is pretty nice and is an alternative to having to support interpolation.

**Comment by jbardin** [comment]
That looks like it's working as intended ;) The only caveat is that some users expect to be able to change the backend "on the fly", rather than running `init` with new parameters each time. 

We may not even be able to do much better conceptually than this, as adding interpolated variables to the backend opens up a whole slew of complexity, but at a minimum this process should get smoother and better documented.
